RenderStatePackager/src/RenderStatePackager.cpp

```diff
@@ -33,217 +33,314 @@ namespace Diligent
                                          RefCntAutoPtr<IShaderSourceInputStreamFactory> pStreamFactory,
+                                         RefCntAutoPtr<IThreadPool>                     pThreadPool,
                                          ARCHIVE_DEVICE_DATA_FLAGS                      DeviceBits) :
     m_pDevice{pDevice},
     m_pStreamFactory{pStreamFactory},
+    m_pThreadPool{pThreadPool},
     m_DeviceBits{DeviceBits}
 {
 }
 
-bool RenderStatePackager::Execute(IArchiver* pArchive, const IRenderStateNotationParser* pNotationParser)
+bool RenderStatePackager::ParseFiles(std::vector<std::string> const& DRSNPaths)
 {
-    DEV_CHECK_ERR(pArchive != nullptr, "pArchive must not be null");
-    DEV_CHECK_ERR(pNotationParser != nullptr, "pNotationParser must not be null");
-    if (pArchive == nullptr || pNotationParser == nullptr)
-        return false;
+    DEV_CHECK_ERR(!DRSNPaths.empty(), "DRSNPaths must not be empty");
 
     try
     {
-        DynamicLinearAllocator Allocator{DefaultRawMemoryAllocator::GetAllocator()};
-
-        auto const& ParserInfo = pNotationParser->GetInfo();
+        std::vector<std::exception_ptr>        Exceptions{};
+        std::vector<RefCntAutoPtr<IAsyncTask>> Tasks{};
 
-        for (Uint32 ShaderID = 0; ShaderID < ParserInfo.ShaderCount; ShaderID++)
+        for (auto const& Path : DRSNPaths)
         {
-            ShaderCreateInfo ShaderCI           = *pNotationParser->GetShaderByIndex(ShaderID);
-            ShaderCI.pShaderSourceStreamFactory = m_pStreamFactory;
-
-            RefCntAutoPtr<IShader> pShader;
-            m_pDevice->CreateShader(ShaderCI, m_DeviceBits, &pShader);
-            if (!pShader)
-                LOG_ERROR_AND_THROW("Failed to create shader from file '", ShaderCI.FilePath, "'.");
-            m_Shaders.emplace(ShaderCI.Desc.Name, pShader);
-        }
+            auto pTask = EnqueueAsyncWork(m_pThreadPool, [Path, this, &Exceptions](Uint32 ThreadId) {
+                try
+                {
+                    RefCntAutoPtr<IRenderStateNotationParser> pDescriptorParser;
+                    CreateRenderStateNotationParserFromFile(Path.c_str(), &pDescriptorParser);
+                    if (!pDescriptorParser)
+                        LOG_ERROR_AND_THROW("Failed to parse file '", Path.c_str(), "'.");
 
-        for (Uint32 RenderPassID = 0; RenderPassID < ParserInfo.RenderPassCount; RenderPassID++)
-        {
-            RenderPassDesc             RPDesc = *pNotationParser->GetRenderPassByIndex(RenderPassID);
-            RefCntAutoPtr<IRenderPass> pRenderPass;
-            m_pDevice->CreateRenderPass(RPDesc, &pRenderPass);
-            if (!pRenderPass)
-                LOG_ERROR_AND_THROW("Failed to create render pass '", RPDesc.Name, "'.");
-            m_RenderPasses.emplace(RPDesc.Name, pRenderPass);
-        }
+                    std::unique_lock<std::mutex> Lock{m_LockRSNParsers};
+                    m_RSNParsers.push_back(pDescriptorParser);
```

Comment:
You have one parser per task. You can resize the m_RSNParsers vector before the loop and the write
`m_RSNParsers[TaskId] = pDescriptorParser;`
without the mutex.